include/deal.II/lac/la_parallel_vector.templates.h

```diff
@@ -79,7 +79,288 @@ namespace LinearAlgebra
+              AssertThrowMPI(ierr);
+
+              ierr = MPI_Info_set(info, "alloc_shared_noncontig", "true");
+              AssertThrowMPI(ierr);
+
+              const std::size_t align_by = 64;
+
+              std::size_t s =
+                ((new_alloc_size * sizeof(Number) + align_by - 1) /
+                 sizeof(Number)) *
+                sizeof(Number);
+
+              ierr = MPI_Win_allocate_shared(
+                s, sizeof(Number), info, comm_shared, &data_this, &mpi_window);
+              AssertThrowMPI(ierr);
 
+              for (unsigned int i = 0; i < size_sm; ++i)
+                {
+                  int        disp_unit;
+                  MPI_Aint   ssize;
+                  const auto ierr = MPI_Win_shared_query(
+                    mpi_window, i, &ssize, &disp_unit, &others[i]);
+                  AssertThrowMPI(ierr);
+                }
+
+              Number *ptr_unaligned = others[rank_sm];
+              Number *ptr_aligned   = ptr_unaligned;
+
+              AssertThrow(std::align(align_by,
+                                     new_alloc_size * sizeof(Number),
+                                     reinterpret_cast<void *&>(ptr_aligned),
+                                     s) != nullptr,
+                          ExcNotImplemented());
+
+              unsigned int n_align_local = ptr_aligned - ptr_unaligned;
+              std::vector<unsigned int> n_align_sm(size_sm);
+
+              ierr = MPI_Allgather(&n_align_local,
+                                   1,
+                                   MPI_UNSIGNED,
+                                   n_align_sm.data(),
+                                   1,
+                                   MPI_UNSIGNED,
+                                   comm_shared);
+              AssertThrowMPI(ierr);
+
+              for (unsigned int i = 0; i < size_sm; ++i)
+                others[i] += n_align_sm[i];
+
+              std::vector<unsigned int> new_alloc_sizes(size_sm);
+
+              ierr = MPI_Allgather(&new_alloc_size,
+                                   1,
+                                   MPI_UNSIGNED,
+                                   new_alloc_sizes.data(),
+                                   1,
+                                   MPI_UNSIGNED,
+                                   comm_shared);
+              AssertThrowMPI(ierr);
+
+              data.values_sm.resize(size_sm);
+              for (unsigned int i = 0; i < size_sm; ++i)
+                data.values_sm[i] =
+                  ArrayView<const Number>(others[i], new_alloc_sizes[i]);
+
+              data.values =
+                Kokkos::View<Number *,
+                             MemorySpaceType,
+                             Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
+                  ptr_aligned, new_alloc_size);
+
+              // Kokkos will not free the memory because the memory is
+              // unmanaged. Instead we use a unique pointer to take care of
```

Comment:
Isn't `value_sm_ptr` a `std::shared_ptr`?

```suggestion
              // unmanaged. Instead, we use a shared pointer to take care of
```